File: lib/features/feed/domain/models/post.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:oasis/features/feed/domain/models/enhanced_poll.dart';

part 'post.freezed.dart';
part 'post.g.dart';

@freezed
abstract class Post with _$Post {
  const factory Post({
    @Default('') String id,
    @JsonKey(name: 'userId') @Default('') String userId,
    @Default('user') String username,
    @JsonKey(name: 'userAvatar') @Default('') String userAvatar,
    String? content,
    @JsonKey(name: 'image_url') String? imageUrl,
    @JsonKey(name: 'thumbnail_url') String? thumbnailUrl,
    @JsonKey(name: 'dominant_color') String? dominantColor,
    @Default([]) @JsonKey(name: 'media_urls') List<String> mediaUrls,
    @Default([]) @JsonKey(name: 'media_types') List<String> mediaTypes,
    @Default([]) List<String> hashtags,
    @Default(false) @JsonKey(name: 'is_spoiler') bool isSpoiler,
    @JsonKey(name: 'community_id') String? communityId,
    @JsonKey(name: 'community_name') String? communityName,
    @JsonKey(name: 'circle_id') String? circleId,
    required DateTime timestamp,
    @Default(0) int likes,
    @Default(0) int comments,
    @Default(0) int shares,
    @Default(false) @JsonKey(name: 'isLiked') bool isLiked,
    @Default(false) @JsonKey(name: 'isBookmarked') bool isBookmarked,
    @Default(false) @JsonKey(name: 'isAd') bool isAd,
    @Default(false) @JsonKey(name: 'isVerified') bool isVerified,
    @JsonKey(name: 'storage_provider') String? storageProvider,
    String? mood,
    EnhancedPoll? poll,
    @Default([]) List<Map<String, dynamic>> collaborators,
  }) = _Post;

  const Post._();

  factory Post.fromJson(Map<String, dynamic> json) =>
      _$PostFromJson(_normalizePostJson(json));

  static Map<String, dynamic> _normalizePostJson(Map<String, dynamic> json) {
    final Map<String, dynamic> normalized = Map.from(json);

    // Handle nested profile data if present (e.g. from joined selects)
    final profile = json['profiles'] ?? json['user'];
    if (profile != null && profile is Map<String, dynamic>) {
      normalized['username'] =
          profile['username'] ?? profile['full_name'] ?? normalized['username'];
      normalized['userAvatar'] =
          profile['avatar_url'] ??
          profile['user_avatar'] ??
          normalized['user_avatar'] ??
          normalized['userAvatar'];
      normalized['isVerified'] =
          profile['is_verified'] ??
          normalized['is_verified'] ??
          normalized['isVerified'];
    }

    normalized['userId'] = json['user_id'] ?? json['userId'];
    normalized['username'] =
        normalized['username'] ?? json['username'] ?? json['full_name'] ?? '';
    normalized['userAvatar'] =
        normalized['userAvatar'] ??
        json['user_avatar'] ??
        json['avatar_url'] ??
        json['userAvatar'] ??
        '';
    normalized['image_url'] = json['image_url'] ?? json['imageUrl'];
    normalized['likes'] = json['likes_count'] ?? json['likes'] ?? 0;
    normalized['comments'] = json['comments_count'] ?? json['comments'] ?? 0;
    normalized['shares'] = json['shares_count'] ?? json['shares'] ?? 0;
    normalized['storage_provider'] =
        json['storage_provider'] ?? json['storageProvider'];
    normalized['timestamp'] =
        json['created_at'] ??
        json['timestamp'] ??
        DateTime.now().toIso8601String();
    normalized['isLiked'] = json['is_liked'] ?? json['isLiked'] ?? false;
    normalized['isBookmarked'] =
        json['is_bookmarked'] ?? json['isBookmarked'] ?? false;
    normalized['is_ad'] = json['is_ad'] ?? json['isAd'] ?? false;
    normalized['isVerified'] =
        normalized['isVerified'] ??
        json['is_verified'] ??
        json['isVerified'] ??
        false;
    normalized['circle_id'] = json['circle_id'] ?? json['circleId'];

    // Handle nested community data
    final community = json['communities'];
    if (community != null && community is Map<String, dynamic>) {
      normalized['community_name'] =
          community['name'] ?? normalized['community_name'];
    }

    // Handle nested poll data from Supabase
    if (json['polls'] != null) {
      final pollsList = json['polls'] as List;
      if (pollsList.isNotEmpty) {
        normalized['poll'] = pollsList.first;
      }
    } else if (json['poll'] != null) {
      normalized['poll'] = json['poll'];
    }

    if (json['collaborators'] != null) {
      final List<dynamic> collabList = json['collaborators'] as List;
      normalized['collaborators'] = collabList.map((c) {
        final Map<String, dynamic> collab = Map.from(c);
        if (c['profiles'] != null) {
          final profile = c['profiles'];
          collab['username'] = profile['username'];
          collab['user_avatar'] = profile['avatar_url'];
          collab['is_verified'] = profile['is_verified'];
        }
        return collab;
      }).toList();
    }

    return normalized;
  }
}

File: winui/src/Oasis.WinUI/Core/Models/Post.cs
using System.Text.Json;

namespace Oasis.WinUI.Core.Models;

/// <summary>
/// A feed post. Field names mirror the Flutter Post model
/// (lib/features/feed/domain/models/post.dart).
/// </summary>
public sealed class Post
{
    public string Id { get; set; } = "";
    public string UserId { get; set; } = "";
    public string Username { get; set; } = "";
    public string UserAvatar { get; set; } = "";
    public string Content { get; set; } = "";
    public string? ImageUrl { get; set; }
    public string? ThumbnailUrl { get; set; }
    public string? DominantColor { get; set; }
    public List<string> MediaUrls { get; set; } = new();
    public List<string> MediaTypes { get; set; } = new();
    public List<string> Hashtags { get; set; } = new();
    public bool IsSpoiler { get; set; }
    public string? CommunityId { get; set; }
    public string? CommunityName { get; set; }
    public string? CircleId { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.Now;
    public long Likes { get; set; }
    public long Comments { get; set; }
    public long Shares { get; set; }
    public bool IsLiked { get; set; }
    public bool IsBookmarked { get; set; }
    public bool IsAd { get; set; }
    public bool IsVerified { get; set; }
    public string? StorageProvider { get; set; }
    public string? Mood { get; set; }
    public string? Poll { get; set; }
    public List<string> Collaborators { get; set; } = new();

    public static Post FromJson(JsonElement el)
    {
        var post = new Post
        {
            Id = JsonUtil.S(el, "id") ?? "",
            UserId = JsonUtil.S(el, "user_id") ?? JsonUtil.S(el, "userId") ?? "",
            Username = JsonUtil.S(el, "username", "full_name")
                       ?? JsonUtil.NestedProfile(el, "profiles", "username")
                       ?? JsonUtil.NestedProfile(el, "user", "username") ?? "",
            UserAvatar = JsonUtil.S(el, "avatar_url", "user_avatar", "userAvatar")
                         ?? JsonUtil.NestedProfile(el, "profiles", "avatar_url") ?? "",
            Content = JsonUtil.S(el, "content") ?? "",
            ImageUrl = JsonUtil.S(el, "image_url", "imageUrl"),
            ThumbnailUrl = JsonUtil.S(el, "thumbnail_url"),
            DominantColor = JsonUtil.S(el, "dominant_color"),
            MediaUrls = JsonUtil.StrList(el, "media_urls", "mediaUrls"),
            MediaTypes = JsonUtil.StrList(el, "media_types", "mediaTypes"),
            Hashtags = JsonUtil.StrList(el, "hashtags"),
            IsSpoiler = JsonUtil.B(el, "is_spoiler"),
            CommunityId = JsonUtil.S(el, "community_id"),
            CommunityName = JsonUtil.NestedProfile(el, "communities", "name")
                            ?? JsonUtil.S(el, "community_name"),
            CircleId = JsonUtil.S(el, "circle_id", "circleId"),
            Timestamp = JsonUtil.Dt(el, "created_at", "timestamp") ?? DateTime.Now,
            Likes = JsonUtil.L(el, "likes_count", "likes"),
            Comments = JsonUtil.L(el, "comments_count", "comments"),
            Shares = JsonUtil.L(el, "shares_count", "shares"),
            IsLiked = JsonUtil.B(el, "is_liked", "isLiked"),
            IsBookmarked = JsonUtil.B(el, "is_bookmarked", "isBookmarked"),
            IsAd = JsonUtil.B(el, "is_ad", "isAd"),
            IsVerified = JsonUtil.B(el, "is_verified", "isVerified"),
            StorageProvider = JsonUtil.S(el, "storage_provider"),
            Mood = JsonUtil.S(el, "mood"),
            Poll = JsonUtil.S(el, "poll"),
            Collaborators = JsonUtil.StrList(el, "collaborators"),
        };

        if (string.IsNullOrEmpty(post.Id) && el.TryGetProperty("id", out var idEl) && idEl.ValueKind != JsonValueKind.String)
            post.Id = idEl.ToString();
        if (post.MediaUrls.Count == 0 && !string.IsNullOrEmpty(post.ImageUrl))
            post.MediaUrls.Add(post.ImageUrl);

        return post;
    }

    public static List<Post> FromJsonArray(string json)
    {
        var result = new List<Post>();
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return result;
            foreach (var el in doc.RootElement.EnumerateArray())
            {
                if (el.ValueKind == JsonValueKind.Object) result.Add(FromJson(el));
            }
        }
        catch
        {
            // malformed payload -> empty feed
        }
        return result;
    }
}
